### DISCUSSION TITLE
Site: Dead link

### DISCUSSION THREAD

**Discussion by za3k** [cc-tweaked/CC-Tweaked#1856] [General]
In getting started, http://www.computercraft.info/forums2/index.php?/topic/15033-computer-basics-i/ is now a dead link.

**Comment by fatboychummy** [comment]
From what I hear, Dan is working on getting a read-only version of the forums back up and running again, but that may take a bit of time.

In the meantime, it may be worthwhile to use an archive link, something like:

https://web.archive.org/web/20230210010158/http://www.computercraft.info/forums2/index.php?/topic/15033-computer-basics-i/

Edit: The Wayback Machine unfortunately didn't properly save the image links on the page.

**Reply by SquidDev** [reply]
Just to add to @fatboychummy's comment, the computercraft.info site is maintained by dan200 (the original author of ComputerCraft). The only website CC:T maintains is tweaked.cc.

While Wayback machine didn't save the image links, they are still available on imgur. Here's an alternative version of the page:

<details><summary>Computer Basics I</summary>

This is a tutorial in the [Ask a Pro Renewal Project](http://www.computercraft.info/forums2/index.php?/topic/14475-ask-a-pro-renewal-project-and-how-you-can-help/) series.

This tutorial has no prerequisite tutorials.

Next suggested tutorial:
[Computer Basics II](http://www.computercraft.info/forums2/index.php?/topic/15041-computer-basics-ii/)

In this tutorial, we will craft our first computer or turtle, explore the command line and lua prompt, create a startup file, and learn the cd and list commands.  To begin, you will need a copy of Minecraft with the appropriate version of ComputerCraft installed.

To craft a computer, you will need seven smooth stone, one redstone dust, and one glass pane.  Arrange these items in the crafting grid like so:


![](http://i.imgur.com/P0AXAY4.png)


Note that you can replace the smooth stone with gold ingots to craft an advanced computer, which adds a colored screen and mouse support.  Once a computer is crafted, it cannot be upgraded into an advanced computer.  Select your new computer and place it in the world like any other block.  Right-click on any side of the computer block to open the computer's interface.  You should see a black rectangle with a grey (or gold) border with some text in the upper-left corner, like this:


![](http://i.imgur.com/eh2b8V7.png)


This is called the terminal or command line interface.  We can type into this and the computer will receive the keystrokes.  Let's type in `list` and hit the Enter key.  We'll see the four letters appear on the screen as we type them, and when we hit Enter, the computer will try to run the "list" program.  It's a valid program, so we will get to see the output it prints, which looks something like this:


![](http://i.imgur.com/J5nIdYW.png)


The list program shows the files and folders in the current directory.  If we are using an advanced computer, the folders' names will be colored lime green, while the file names are white.  On a normal computer, the files' and folders' names are all in white, of course.  As you can see, we only have a folder named "rom" on this computer.  The rom folder is present on all computers. It is also read-only, which means that you cannot make any changes to any files or folders in it, or add any new files or folders to it.

Let's see what's inside the rom folder.  The first way to figure out what's inside the rom folder is simply to tell the list program to show the contents of that folder instead of the current folder.  We can run "list" again, but this time give it an *argument* telling it a specific folder to show the contents of.  Many of the programs in ComputerCraft accept arguments to change their behavior, and some require them to work at all.  Type in `list rom` to show the contents of the rom folder:


![](http://i.imgur.com/93QEpaD.png)


Now we can see the contents of the rom folder.  But there's another way to get there which can be handy at times.  We can change our current directory with the "cd" (change directory) program. Since we're in the main folder (or *root*) of the computer's drive, we type `cd rom` to change our current directory to the rom folder:


![](http://i.imgur.com/2X41u7t.png)


Now our prompt looks a little different.  It's been hard to tell so far, but the prompt has been showing us our current directory all along.  Now that our current directory is the rom folder, it has changed to "rom>".  Now if we type `list` again, with no arguments, it will show us the current directory's contents.  Now that the current directory is the rom folder, it will show us the same results as when we ran `list rom` earlier.


![](http://i.imgur.com/q9I0CS8.png)


We can return to the main folder of the computer by using the `cd ..` command.  Running `cd ..` always changes the directory to be one level higher than the current directory, if possible.  If our current directory was /rom/programs, running `cd ..` would change our current directory to be /rom.  Running it again would change our current directory to be the main directory.  Let's go ahead and run `cd ..` now:


![](http://i.imgur.com/6i3HZwu.png)


Note that our prompt has gone back to being just ">", since we are back at the root of the drive. Feel free to play around with navigating through the directories a little bit to familiarize yourself with these two commands.  There is one way to get yourself into trouble with the cd command, so I'll demonstrate that problem.  In this example, we've changed our current directory to /rom/help:


![](http://i.imgur.com/RQjcQ7Y.png)


We try to use `cd ..` to get out of this folder to keep exploring, but wait, what?


![](http://i.imgur.com/gYX41nm.png)


An error message?!  Why did the cd program suddenly break?  Well, the answer is that the help directory contains a bunch of plain-text help files to provide information on each of the programs that ComputerCraft comes with.  The shell program that creates the command-line interface that we are using looks for the programs that we tell it to run in the current directory first.  Since there is a file named "cd" in the /rom/help directory, which is our current directory, it tries to run that file.  But that file is just a plain text help file, so when it tries to run it as a program, it just throws an error and stops!

To fix the problem, we have to tell the shell exactly which "cd" program to try to run, so that it won't try to run the help file for cd instead.  The cd program is actually located at /rom/programs/cd, so we type in `/rom/programs/cd ..` to escape the help directory.


![](http://i.imgur.com/pkeXZqy.png)


Let's move on to the Lua prompt.  The "lua" program starts another command-line interface, but this one is a little different than the shell.  In the Lua prompt, we can run little bits of actual lua code directly, instead of running program files with Lua in them like the shell does. Type `cd /` to change to the root directory directly, then type `lua` to start the Lua prompt.


![](http://i.imgur.com/6RGk09N.png)


You should see a screen a little like this one.  Take note of the "Call exit() to exit." line.  We'll run a couple of simple lua lines, then use the exit() function to exit the lua prompt.  Type in `x = 1 + 1` and hit Enter.  The prompt will accept the line and prompt you for another.  Type `x`, and the value of the variable (2) should be printed on the next line, something like this:


![](http://i.imgur.com/dqNu6Qu.png)


Type `exit()` to quit the Lua prompt when you're ready.  We're back to our familiar command-line.  Let's create a startup file next.  When a computer is booted up, it will run the Lua code in its startup file.  This is commonly used to run a program that you want the computer to do each time it starts, like a door lock program.  Type `edit startup` to open the editor.  We're starting up the edit program with "startup" as its argument.  This tells the edit program that we want to edit the file named "startup" in the current directory.  Since that file doesn't exist, it will open up a blank editor.  When we save the file in the editor, it will automatically create the file since it did not exist already.  The edit program is one of the programs in ComputerCraft that requires an argument when you start it so that it knows what file you're wanting to edit.


![](http://i.imgur.com/2N49NQX.png)


We have a blank file, since this is a new computer.  Let's make it display a line of text using the print() function.  We add `print("Hello, World!")` to the file.  To save the file, we hit Ctrl like it says on the bottom line of the screen (on some keyboards, this will be AltGr or another key).  A small menu pops up on the bottom line.  We can either hit Enter or the S key to save the file. The bottom line will change to "Saved as startup" if all went well:


![](http://i.imgur.com/VUPKpB7.png)


Now we hit Ctrl again to bring up the menu, and either hit the E key to exit, or use the right arrow key to select Exit and hit Enter.  We've now created a startup file and are back to the command line prompt.  Let's reboot the computer to see what our startup file will have changed.  Type `reboot` and hit Enter.  The computer will reboot, and when it comes up, there's our printed line!


![](http://i.imgur.com/3CokmkM.png)


In this tutorial, we learned how to create a computer in ComputerCraft.  We learned how to use the shell and lua prompts to run programs or bits of lua code.  We learned how to use two of the programs ComputerCraft comes with, cd and list.  We know how to create, edit and save files in the edit program, and we have learned what the startup file is and does.  Thanks for reading!

</details>

**Reply by SquidDev** [reply]
Lem has provided me a snapshot of the forum taken back in April 2020, which I've put online. This is definitely not perfect, but should at least help until the forums are properly back. You can replace replace `www.computercraft.info` with `ccf.squiddev.cc` in the URL to view the mirror ([example](https://ccf.squiddev.cc/forums2/index.php?/topic/15033-computer-basics-i/)).

**Comment by za3k** [comment]
Aww thanks rumored-dan200!
